<h1>Week 3 Assignment</h1>

<h1>Task 1:</h1>

<h4>Import Numpy and Pandas</h4>

In [12]:
import numpy as np
import pandas as pd

<h4>Read data from the Wiki page using Pandas</h4>

In [13]:
df = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")[0]
print(df.shape)
df.head()

(180, 3)


Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

<h4>Drop rows that don't have Borough</h4>

In [14]:
df.drop(df.index[df['Borough'] == "Not assigned"], inplace = True)
df.reset_index(drop=True, inplace=True)
print(df.shape)
df.head()

(103, 3)


Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

<h4>Check if there are any "Not assigned" Neighbourhood</h4>

In [15]:
print("Number of Not assigned Neighbourhood: " + str(df[df['Neighbourhood'] == "Not assigned"].shape[0]))

Number of Not assigned Neighbourhood: 0


<h4>There is no "Not assigned" Neighbourhood, so we don't need to perform any clean-up</h4>

<h4>Check to see if there are any duplicated Postal Code</h4>

In [16]:
dup = df[df.duplicated(['Postal Code'])]
print(dup)

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []


<h4>There is no duplicated Postal Code so no more clean-up is needed</h4>

<h2>So here is the final number of rows of the data frame:</h2>

In [17]:
df.shape[0]

103

<h1>Task 2:</h1>

<h4>We will use the CSV file for coordinates</h4>

In [18]:
df_latlong = pd.read_csv("https://raw.githubusercontent.com/uclinux83/Coursera_Capstone/master/Geospatial_Coordinates.csv")
df_latlong.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [19]:
df_latlong.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

<h4>Join the 2 dataframes together</h4>

In [20]:
df_new = pd.merge(df, df_latlong, on="Postal Code")
df_new.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

<h4>Check the size of the new dataframe</h4>

In [9]:
df_new.shape

(103, 5)

<h4>It has 103 rows and 5 columns as expected. So we're good for Task 2!</h4>

<h1>Task 3:</h1>

<h4>Install required libraries</h4>

In [1]:
!pip install folium

     |████████████████████████████████| 93 kB 3.2 MB/s  eta 0:00:01


<h4>Import required libraries</h4>

In [66]:
import folium
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
print('Libraries imported.')

Libraries imported.


<h4>Rename the column Neighbourhood to Neighborhood</h4>

In [77]:
df_new.rename(columns={"Neighbourhood": "Neighborhood"}, inplace = True)
df_new.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

<h4>Get the location of Toronto</h4>

In [78]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


<h4>Draw on map</h4>

In [80]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_new['Latitude'], df_new['Longitude'], df_new['Borough'], df_new['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<h4>Now we get the list of neighborhood in Downtown Toronto to explore</h4>

In [81]:
downtown_toronto = df_new[df_new['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

In [82]:
address = 'Downtown Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


In [28]:
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(downtown_toronto['Latitude'], downtown_toronto['Longitude'], downtown_toronto['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

<h3>Now just follow the instructions in the lab to explore Downtown Toronto</h3>

In [83]:
CLIENT_ID = 'HIDDEN' #Hide for submission
CLIENT_SECRET = 'HIDDEN' #Hide for submission
VERSION = '20201220' 
LIMIT = 100 

In [84]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [85]:
downtown_venues = getNearbyVenues(names=downtown_toronto['Neighborhood'],
                                   latitudes=downtown_toronto['Latitude'],
                                   longitudes=downtown_toronto['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


In [86]:
print(downtown_venues.shape)
downtown_venues.head()

(1227, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [87]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     61   
Christie                                                               15   
Church and Wellesley                                                   80   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              62   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               96   
Rosedale                                                                4   
St. James Town                                                         81   
St. James Town, Cabbagetown                                            42   
Stn A PO Boxes                                                         98   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         33   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      61   
Christie                                                                15   
Church and Wellesley                                                    80   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               62   
Queen's Park, Ontario Provincial Government                             33   
Regent Park, Harbourfront                                               46   
Richmond, Adelaide, King                                                96   
Rosedale                                                                 4   
St. James Town                                                          81   
St. James Town, Cabbagetown                                             42   
Stn A PO Boxes                                                          98   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          33   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            59              59   
CN Tower, King and Spadina, Railway Lands, Harb...     17              17   
Central Bay Street                                     61              61   
Christie                                               15              15   
Church and Wellesley                                   80              80   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

In [88]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 210 uniques categories.


In [89]:
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood'] 
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Tea Room  Thai Restaurant  \
0                 0                    0  ...         0                0   
1                 0                    0  ...         0                0   
2                 0                    0  ...         0                0   
3                 0                    0  ...         0                0   
4                 0                    0  ...         0                0   

   Theater  Theme Restaurant  Trail  Train Station  \
0        0                 0      0              0   
1        0                 0      0              0   
2        0                 0      0              0   
3        0                 0      0              0   
4        0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 210 columns]

In [90]:
downtown_onehot.shape

(1227, 210)

In [91]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                  Central Bay Street     0.016393   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.025000   
5                      Commerce Court, Victoria Hotel     0.000000   
6              First Canadian Place, Underground city     0.000000   
7                            Garden District, Ryerson     0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000   
10        Queen's Park, Ontario Provincial Government     0.030303   
11                          Regent Park, Harbourfront     0.021739   
12                           Richmond, Adelaide, King     0.000000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                        St. James Town, Cabbagetown     0.000000   
16                                     Stn A PO Boxes     0.010204   
17           Toronto Dominion Centre, Design Exchange     0.000000   
18                     University of Toronto, Harbord     0.030303   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0           0.0000             0.0000  0.000000            0.000000   
1           0.0000             0.0000  0.058824            0.058824   
2           0.0000             0.0000  0.000000            0.000000   
3           0.0000             0.0000  0.000000            0.000000   
4           0.0125             0.0125  0.000000            0.000000   
5           0.0000             0.0000  0.000000            0.000000   
6           0.0000             0.0000  0.000000            0.000000   
7           0.0000             0.0000  0.000000            0.000000   
8           0.0000             0.0000  0.000000            0.000000   
9           0.0000             0.0000  0.000000            0.000000   
10          0.0000             0.0000  0.000000            0.000000   
11          0.0000             0.0000  0.000000            0.000000   
12          0.0000             0.0000  0.000000            0.000000   
13          0.0000             0.0000  0.000000            0.000000   
14          0.0000             0.0000  0.000000            0.000000   
15          0.0000             0.0000  0.000000            0.000000   
16          0.0000             0.0000  0.000000            0.000000   
17          0.0000             0.0000  0.000000            0.000000   
18          0.0000             0.0000  0.000000            0.000000   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0       0.000000        0.000000         0.000000          0.000000  ...   
1       0.058824        0.117647         0.176471          0.117647  ...   
2       0.000000        0.000000         0.000000          0.000000  ...   
3       0.000000        0.000000         0.000000          0.000000  ...   
4       0.000000        0.000000         0.000000          0.000000  ...   
5       0.000000        0.000000         0.000000          0.000000  ...   
6       0.000000        0.000000         0.000000          0.000000  ...   
7       0.000000        0.000000         0.000000          0.000000  ...   
8       0.000000        0.000000         0.000000          0.000000  ...   
9       0.000000        0.000000         0.000000          0.000000  ...   
10      0.000000        0.000000         0.000000          0.000000  ...   
11      0.000000        0.000000         0.000000          0.000000  ...   
12      0.000000        0.000000         0.000000          0.000000  ...   
13      0.000000        0.000000         0.000000          0.000000  ...   
14      0.000000        0.000000         0.000000          0.000000  ...   
15 

In [92]:
downtown_grouped.shape

(19, 210)

In [93]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [100]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                Bakery    Seafood Restaurant   
1        Airport Lounge      Airport Terminal       Harbor / Marina   
2        Sandwich Place    Italian Restaurant                  Café   
3                  Café                  Park            Baby Store   
4      Sushi Restaurant   Japanese Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Farmers Market              Beer Bar              Pharmacy   
1         Boat or Ferry               Airport    Airport Food Court   
2           Salad Place       Bubble Tea Shop          Burger Joint   
3    Italian Restaurant             Nightclub           Candy Store   
4               Gay Bar           Yoga Studio                   Pub   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0            Restaurant                Cheese Shop              Irish Pub  
1          Airport Gate           Sculpture Garden                  Plane  
2       Thai Restaurant  Middle Eastern Restaurant      Indian Restaurant  
3            Restaurant                Coffee Shop       Airport Terminal  
4           Men's Store   Mediterranean Restaurant                  Hotel

<h2>Done exploring. Now we do clustering</h2>

In [101]:
kclusters = 5
downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)
# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 0, 1, 2, 1, 1, 1, 1, 1, 4], dtype=int32)

In [102]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = downtown_toronto

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

downtown_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               1           Coffee Shop   
2  43.657162 -79.378937               1           Coffee Shop   
3  43.651494 -79.375418               1           Coffee Shop   
4  43.644771 -79.373306               1           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                  Café                Bakery   
1      Sushi Restaurant           Yoga Studio                  Café   
2        Clothing Store                  Café        Cosmetics Shop   
3                  Café        Cosmetics Shop   American Restaurant   
4          Cocktail Bar                Bakery    Seafood Restaurant   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0                   Pub             Breakfast Spot               Theater   
1                   Bar                   Beer Bar    Italian Restaurant   
2       Bubble Tea Shop  Middle Eastern Restaurant   Japanese Restaurant   
3          Cocktail Bar        Japanese Restaurant                 Hotel   
4        Farmers Market                   Beer Bar              Pharmacy   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Yoga Studio        Farmers Market             Shoe Store  
1   Japanese Restaurant            Restaurant                  Diner  
2                 Hotel             Bookstore       Ramen Restaurant  
3        Clothing Store           Cheese Shop                    Gym  
4            Restaurant           Cheese Shop              Irish Pub

<h4>Now we show the map</h4>

In [103]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighborhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<h3>Now we look at the clusters:</h3>

In [104]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               0       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13        Airport Lounge      Airport Terminal       Harbor / Marina   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13         Boat or Ferry               Airport    Airport Food Court   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13          Airport Gate      Sculpture Garden                  Plane

In [105]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               1           Coffee Shop   
1   Downtown Toronto               1           Coffee Shop   
2   Downtown Toronto               1           Coffee Shop   
3   Downtown Toronto               1           Coffee Shop   
4   Downtown Toronto               1           Coffee Shop   
5   Downtown Toronto               1           Coffee Shop   
7   Downtown Toronto               1           Coffee Shop   
8   Downtown Toronto               1           Coffee Shop   
9   Downtown Toronto               1           Coffee Shop   
10  Downtown Toronto               1           Coffee Shop   
15  Downtown Toronto               1           Coffee Shop   
16  Downtown Toronto               1           Coffee Shop   
17  Downtown Toronto               1           Coffee Shop   
18  Downtown Toronto               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park                  Café                Bakery   
1       Sushi Restaurant           Yoga Studio                  Café   
2         Clothing Store                  Café        Cosmetics Shop   
3                   Café        Cosmetics Shop   American Restaurant   
4           Cocktail Bar                Bakery    Seafood Restaurant   
5         Sandwich Place    Italian Restaurant                  Café   
7                   Café            Restaurant         Deli / Bodega   
8               Aquarium                 Hotel                  Café   
9                  Hotel                  Café            Restaurant   
10            Restaurant                  Café                 Hotel   
15    Italian Restaurant    Seafood Restaurant            Restaurant   
16                  Café            Restaurant                   Pub   
17                  Café                 Hotel   Japanese Restaurant   
18      Sushi Restaurant   Japanese Restaurant            Restaurant   

   5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0                    Pub             Breakfast Spot               Theater   
1                    Bar                   Beer Bar    Italian Restaurant   
2        Bubble Tea Shop  Middle Eastern Restaurant   Japanese Restaurant   
3           Cocktail Bar        Japanese Restaurant                 Hotel   
4         Farmers Market                   Beer Bar              Pharmacy   
5            Salad Place            Bubble Tea Shop          Burger Joint   
7                    Gym            Thai Restaurant                 Hotel   
8    Fried Chicken Joint                 Restaurant        Scenic Lookout   
9    Japanese Restaurant         Italian Restaurant   American Restaurant   
10    Italian Restaurant                        Gym         Deli / Bodega   
15                Bakery        Japanese Restaurant              Beer Bar   
16                Bakery         Italian Restaurant           Pizza Place   
17            Restaurant                        Gym      Asian Restaurant   
18               Gay Bar                Yoga Studio                   Pub   

   8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0            Yoga Studio             Farmers Market             Shoe Store  
1    Japanese Restaurant                 Restaurant                  Diner  
2                  Hotel                  Bookstore       Ramen Restaurant  
3         Clothing Store                Cheese Shop                    Gym  
4             Restaurant                Cheese Shop              Irish Pub  
5        Thai Restaurant  Middle Eastern Restaurant      Indian Restaurant  
7          Burrito Place           Sushi Restaurant                 Bakery  
8                Brewery         Italian Restaurant            Music Venue  
9            Salad Place         Seafood Restaurant       Sushi Restaurant  
10    Seafood Restaurant        Japanese Restaurant    American Restaurant 

In [106]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6  Downtown Toronto               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                  Café                  Park            Baby Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6    Italian Restaurant             Nightclub           Candy Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6            Restaurant           Coffee Shop       Airport Terminal

In [107]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                 Trail            Playground              Wine Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14          Cupcake Shop            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14               Dog Run   Distribution Center         Discount Store

In [108]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
11  Downtown Toronto               4                  Café   
12  Downtown Toronto               4                  Café   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11                      Bookstore                   Bar    Italian Restaurant   
12  Vegetarian / Vegan Restaurant           Coffee Shop    Mexican Restaurant   

    5th Most Common Venue 6th Most Common Venue    7th Most Common Venue  \
11    Japanese Restaurant                Bakery              Yoga Studio   
12  Vietnamese Restaurant          Dessert Shop  Comfort Food Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11              Beer Bar           College Gym         Sandwich Place  
12  Caribbean Restaurant         Grocery Store                   Park

<h1>That's it! Thank you for reviewing!</h1>